File: execution/packages/execution-gateway-protocol/src/protocol.ts
import { bool, bytes, identity, invalid, jsonValue, messageText, object, text, uuid } from "./validation.ts";
import type { Json } from "./validation.ts";
import { digest } from "./crypto.ts";
export const PROTOCOL_VERSION = 1;
export const MAX_NATIVE_BYTES = 8 * 1024 * 1024;
export const MAX_HTTP_BYTES = MAX_NATIVE_BYTES + 64 * 1024;
export const MAX_FRAME_BYTES = MAX_NATIVE_BYTES + 128 * 1024;
export const OPERATIONS = ["request.cancel", "runtime.capabilities",
  "execution.exec", "execution.interact", "execution.close", "filesystem.read", "filesystem.write", "filesystem.patch",
  "repl.execute", "repl.collect", "repl.interrupt", "repl.reset", "repl.close"] as const;
export type OperationName = typeof OPERATIONS[number];
export function operationName(value: unknown): OperationName {
  if (typeof value !== "string" || !(OPERATIONS as readonly string[]).includes(value)) throw invalid("Unsupported native operation.");
  return value as OperationName;
}
export interface Callback { receiver: string; context: Json }
export function callback(value: unknown): Callback {
  const o = object(value, ["receiver", "context"]);
  const context = o.context === undefined ? null : jsonValue(o.context);
  if (bytes(JSON.stringify(context)) > 8192) throw invalid("Callback context exceeds 8 KiB.");
  return { receiver: identity(o.receiver), context };
}
export interface Submission {
  requestId: string; runtimeGeneration: string;
  operation: { operation: OperationName; params: Json };
  callback: Callback;
}
export function submission(value: unknown): Submission {
  const o = object(value, ["requestId", "runtimeGeneration", "operation", "callback"]), op = object(o.operation, ["operation", "params"]);
  const params = jsonValue(object(op.params));
  if (bytes(JSON.stringify(params)) > MAX_NATIVE_BYTES) throw invalid("Native parameters exceed limit.");
  return { requestId: identity(o.requestId), runtimeGeneration: uuid(o.runtimeGeneration),
    operation: { operation: operationName(op.operation), params }, callback: callback(o.callback) };
}
export interface RoutingEnvelope {
  machineId: string; runtimeGeneration: string; requestId: string; requestHash: string; callback: Callback;
}
export function routingFields(value: unknown): RoutingEnvelope {
  const o = object(value);
  return { machineId: uuid(o.machineId), runtimeGeneration: uuid(o.runtimeGeneration), requestId: identity(o.requestId),
    requestHash: digest(o.requestHash), callback: callback(o.callback) };
}
export interface Hello {
  type: "hello"; protocolVersion: 1; runtimeGeneration: string; daemonVersion: string; os: string; arch: string; operations: OperationName[];
}
export function hello(value: unknown): Hello {
  const o = object(value, ["type", "protocolVersion", "runtimeGeneration", "daemonVersion", "os", "arch", "operations"]);
  if (o.type !== "hello" || o.protocolVersion !== PROTOCOL_VERSION || !Array.isArray(o.operations) || o.operations.length > OPERATIONS.length) throw invalid("Unsupported daemon hello.");
  const operations = o.operations.map(operationName); if (new Set(operations).size !== operations.length) throw invalid("Duplicate capability.");
  return { type: "hello", protocolVersion: 1, runtimeGeneration: uuid(o.runtimeGeneration), daemonVersion: text(o.daemonVersion, 128), os: text(o.os, 64), arch: text(o.arch, 64), operations };
}
export type Outcome = { status: "ok"; result: Json } | { status: "error"; error: { code: string; message: string; uncertain: boolean } };
export function outcome(value: unknown): Outcome {
  const o = object(value);
  if (o.status === "ok") { object(o, ["status", "result"]); return { status: "ok", result: jsonValue(o.result) }; }
  if (o.status === "error") { object(o, ["status", "error"]); const e = object(o.error, ["code", "message", "uncertain"]); return { status: "error", error: { code: identity(e.code), message: messageText(e.message, 8192), uncertain: bool(e.uncertain) } }; }
  throw invalid("Invalid native outcome.");
}
export interface CompletionEvent extends RoutingEnvelope { protocolVersion: 1; deliveryId: string; resultHash: string; outcome: Outcome }
export interface CompletionReceipt { status: "accepted"; deliveryId: string; requestId: string; requestHash: string; resultHash: string }
export type CompletionReply = CompletionReceipt | { status: "rejected"; code: string; retryable: boolean };
export interface CompletionReceiver { acceptExecutionResult(event: CompletionEvent): Promise<CompletionReply> }
export type SecretKind = "daemon" | "execution";
export function secretKind(value: unknown): SecretKind {
  if (value !== "daemon" && value !== "execution") throw invalid("Expected daemon or execution secret.");
  return value;
}
export interface MachineView {
  machineId: string; name: string; deleted: boolean; daemonSecretVersion: number; executionSecretVersion: number;
  connectionEpoch: number; createdAt: number; lastConnectedAt: number | null; lastDisconnectedAt: number | null;
  connectionStatus: "offline" | "connecting" | "ready"; runtimeGeneration: string | null; lastSeenAt: number | null;
  daemon: Omit<Hello, "type" | "protocolVersion" | "runtimeGeneration"> | null;
}

File: execution/apps/process-execution-daemon/src/protocol.rs
use crate::Result;
use process_execution_core::{Operation, ProcessExecutionCore, Request};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use uuid::Uuid;
pub const MAX_RESULT: usize = 8 * 1024 * 1024;
pub const MAX_FRAME: usize = MAX_RESULT + 128 * 1024;
pub const OPERATIONS: &[&str] = &[
    "request.cancel",
    "runtime.capabilities",
    "execution.exec",
    "execution.interact",
    "execution.close",
    "filesystem.read",
    "filesystem.write",
    "filesystem.patch",
    "repl.execute",
    "repl.collect",
    "repl.interrupt",
    "repl.reset",
    "repl.close",
];
#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Incoming {
    #[serde(rename = "type")]
    pub kind: String,
    pub protocol_version: u8,
    pub dispatch_id: Uuid,
    pub request_id: String,
    pub request_hash: String,
    pub runtime_generation: Uuid,
    pub operation: Value,
    pub routing_envelope: String,
}
pub enum Action {
    Cancel(String),
    Capabilities,
    Native(Box<Operation>),
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Target {
    request_id: String,
}
pub fn identity(s: &str) -> bool {
    !s.is_empty()
        && s.len() <= 128
        && s.as_bytes()[0].is_ascii_alphanumeric()
        && s.bytes()
            .all(|c| c.is_ascii_alphanumeric() || b"_.:@-".contains(&c))
}
pub fn digest(s: &str) -> bool {
    s.len() == 64
        && s.bytes()
            .all(|c| c.is_ascii_digit() || (b'a'..=b'f').contains(&c))
}
pub fn hash(value: &Value) -> Result<String> {
    Ok(format!(
        "{:x}",
        Sha256::digest(serde_json_canonicalizer::to_vec(value)?)
    ))
}
impl Incoming {
    pub fn key(&self) -> String {
        format!("{}:{}", self.runtime_generation, self.request_id)
    }
    pub fn action(&self) -> Result<Action> {
        if self.kind != "request"
            || self.protocol_version != 1
            || !identity(&self.request_id)
            || !digest(&self.request_hash)
            || self.routing_envelope.len() > 32768
            || self.routing_envelope.is_empty()
        {
            return Err("invalid request envelope".into());
        }
        let op = self
            .operation
            .as_object()
            .ok_or("operation must be an object")?;
        if op.len() != 2 {
            return Err("operation requires only operation and params".into());
        }
        let name = op
            .get("operation")
            .and_then(Value::as_str)
            .ok_or("missing operation")?;
        let params = op
            .get("params")
            .filter(|v| v.is_object())
            .ok_or("params must be an object")?
            .clone();
        let empty = || -> Result<()> {
            if params.as_object().unwrap().is_empty() {
                Ok(())
            } else {
                Err("operation takes no params".into())
            }
        };
        Ok(match name {
            "runtime.capabilities" => {
                empty()?;
                Action::Capabilities
            }
            "request.cancel" => {
                let id = serde_json::from_value::<Target>(params)?.request_id;
                if !identity(&id) {
                    return Err("invalid target request_id".into());
                }
                Action::Cancel(id)
            }
            _ => Action::Native(Box::new(serde_json::from_value(self.operation.clone())?)),
        })
    }
    pub fn reply(&self, error: Option<(&str, &str, bool, bool)>) -> Value {
        let mut reply = json!({"type":"accepted","dispatchId":self.dispatch_id,"requestId":self.request_id,"requestHash":self.request_hash,"runtimeGeneration":self.runtime_generation});
        if let Some((code, message, retryable, uncertain)) = error {
            reply["type"] = json!("rejected");
            reply["error"] =
                json!({"code":code,"message":message,"retryable":retryable,"uncertain":uncertain});
        }
        reply
    }
}
pub async fn execute(core: &ProcessExecutionCore, req: &Incoming, action: Action) -> Value {
    let result = match action {
        Action::Capabilities => Ok(core.capabilities()),
        Action::Cancel(id) => core.cancel(&id).map(|()| json!({"state":"cancelled"})),
        Action::Native(operation) => {
            core.execute(Request {
                request_id: req.request_id.clone(),
                operation: *operation,
            })
            .await
        }
    };
    let result = match result {
        Ok(value) => json!({"status":"ok","result":value}),
        Err(e) => {
            let uncertain = matches!(
                e.code,
                process_execution_core::ErrorCode::Io
                    | process_execution_core::ErrorCode::Unavailable
                    | process_execution_core::ErrorCode::ResourceLimit
                    | process_execution_core::ErrorCode::Cancelled
            );
            json!({"status":"error","error":{"code":e.code,"message":e.message.chars().take(4096).collect::<String>(),"uncertain":uncertain}})
        }
    };
    if serde_json::to_vec(&result).map_or(true, |v| v.len() > MAX_RESULT) {
        error(
            "RESULT_TOO_LARGE",
            "Native result exceeded the transport limit; effects may have occurred.",
            true,
        )
    } else {
        result
    }
}
pub fn error(code: &str, message: &str, uncertain: bool) -> Value {
    json!({"status":"error","error":{"code":code,"message":message,"uncertain":uncertain}})
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn canonical_json_matches_javascript_numbers_and_utf16_key_order() {
        let value = json!({"z":-0.0,"a":1.0,"b":1e-7,"c":1e21,"d":9007199254740993u64,"\u{e000}":1,"\u{10000}":2});
        assert_eq!(
            serde_json_canonicalizer::to_string(&value).unwrap(),
            "{\"a\":1,\"b\":1e-7,\"c\":1e+21,\"d\":9007199254740992,\"z\":0,\"𐀀\":2,\"\":1}"
        );
    }
}
